radix-engine/src/blueprints/locker/blueprint.rs

```diff
@@ -0,0 +1,764 @@
+    {
+        // Recover the resources from the vault.
+        let bucket = Self::with_vault_create_on_traversal(
+            claimant.0,
+            resource_address,
+            api,
+            |mut vault, api| vault.take(amount, api),
+        )??;
+
+        // Emitting the event
+        let (resource_address, resource_specifier) = bucket_to_resource_specifier(&bucket, api)?;
+        Runtime::emit_event(
+            api,
+            RecoveryEvent {
+                claimant,
+                resource_address,
+                resources: resource_specifier,
+            },
+        )?;
+
+        // Return
+        Ok(bucket)
+    }
+
+    fn recover_non_fungibles<Y>(
+        AccountLockerRecoverNonFungiblesInput {
+            claimant,
+            resource_address,
+            ids,
+        }: AccountLockerRecoverNonFungiblesInput,
+        api: &mut Y,
+    ) -> Result<AccountLockerRecoverNonFungiblesOutput, RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        // Recover the resources from the vault.
+        let bucket = Self::with_vault_create_on_traversal(
+            claimant.0,
+            resource_address,
+            api,
+            |mut vault, api| vault.take_non_fungibles(ids, api),
+        )??;
+
+        // Emitting the event
+        let (resource_address, resource_specifier) = bucket_to_resource_specifier(&bucket, api)?;
+        Runtime::emit_event(
+            api,
+            RecoveryEvent {
+                claimant,
+                resource_address,
+                resources: resource_specifier,
+            },
+        )?;
+
+        // Return
+        Ok(bucket)
+    }
+
+    fn claim<Y>(
+        AccountLockerClaimInput {
+            claimant,
+            resource_address,
+            amount,
+        }: AccountLockerClaimInput,
+        api: &mut Y,
+    ) -> Result<AccountLockerClaimOutput, RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        // Read and assert against the owner role of the claimant.
+        let claimant_owner_role = api
+            .call_module_method(
+                claimant.0.as_node_id(),
+                AttachedModuleId::RoleAssignment,
+                ROLE_ASSIGNMENT_GET_OWNER_ROLE_IDENT,
+                scrypto_encode(&RoleAssignmentGetOwnerRoleInput).unwrap(),
+            )
+            .map(|rtn| scrypto_decode::<RoleAssignmentGetOwnerRoleOutput>(&rtn).unwrap())?;
+        Runtime::assert_access_rule(claimant_owner_role.rule, api)?;
+
+        // Recover the resources from the vault.
+        let bucket = Self::with_vault_create_on_traversal(
+            claimant.0,
+            resource_address,
+            api,
+            |mut vault, api| vault.take(amount, api),
+        )??;
+
+        // Emitting the event
+        let (resource_address, resource_specifier) = bucket_to_resource_specifier(&bucket, api)?;
+        Runtime::emit_event(
+            api,
+            RecoveryEvent {
```

Comment:
Wrong event emitted

Reply:
Ooof thanks! We have a pretty comprehensive test for state reconcilation from events: `state_of_the_account_locker_can_be_reconciled_from_events_alone`. This was not picked up by the test since Claim and Recover events have the same effect on state.